### PR TITLE
docs(docs-infra): Add support for function overloads

### PR DESCRIPTION
**Pull request by JeanMeche**
This improves of the API docs  by adding support for function overloads:

Example:  https://ng-dev-previews-fw--pr-angular-angular-57255-adev-prev-wntct8vv.web.app/api/core/inject

### CODE REVIEW

File: adev/shared-docs/pipeline/api-gen/rendering/entities.ts [hunk 1]
```diff
@@ -132,15 +132,22 @@ export interface PipeEntry extends ClassEntry {
   // TODO: add `isPure`.
 }
 
-export interface FunctionEntry extends DocEntry {
+export interface FunctionMetadata extends DocEntry {
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/rendering/entities.ts:135]: How about `FunctionSignatureMetadata`?

File: adev/shared-docs/pipeline/api-gen/rendering/shiki/shiki.ts [hunk 2]
```diff
@@ -22,13 +22,21 @@ export async function initHighlighter() {
...
+          dark: 'github-dark',
+        },
+        cssVariablePrefix: '--shiki-',
+        defaultColor: false,
+      })
+      // remove the leading space of the element after the "function" element
+      .replace(/(<[^>]*>function<\/\w+><[^>]*>)(\s)(\w+<\/\w+>)/g, '$1$3')
+      // Shiki requires the keyword function for highlighting functions signatures
+      // We don't want to display it so we remove elements with the keyword
+      .replace(/<[^>]*>function<\/\w+>/g, '')
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/rendering/shiki/shiki.ts:40]: This makes sense, but isn't this helper `codeToHtml` pretty "generic". It does feel wrong to simply remove the keyword here wherever `codeToHtml` is used?

File: adev/shared-docs/pipeline/api-gen/rendering/templates/class-member.tsx [hunk 3]
```diff
@@ -13,49 +13,60 @@ import {
...
-            <RawHtml value={member.htmlDescription} />
-          </div>
-        );
+      {props.members.map((member, index) => {
+        const hasSignature = isClassMethodEntry(member) && member.signatures?.[index] !== undefined;
+
+        if (hasSignature) {
+          const renderableMember = getFunctionMetadataRenderable(member.signatures[index]);
+          return (
+            <ClassMethodInfo entry={renderableMember} isOverloaded={props.members.length > 1} />
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/rendering/templates/class-member.tsx:46]: Should this use `member.signatures.length > 1`?

File: adev/shared-docs/pipeline/api-gen/rendering/templates/class-member.tsx [hunk 4]
```diff
@@ -13,49 +13,60 @@ import {
...
-            {props.members.every((member) => member.deprecationMessage !== null) ? (
-              <DeprecatedLabel entry={props.members[0]} />
-            ) : (
-              <></>
-            )}
-            <RawHtml value={member.htmlDescription} />
-          </div>
-        );
+      {props.members.map((member, index) => {
+        const hasSignature = isClassMethodEntry(member) && member.signatures?.[index] !== undefined;
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/rendering/templates/class-member.tsx:41]: Is this correct? Is the index based on the member list related to an entry in the signatures` field?

File: adev/shared-docs/pipeline/api-gen/rendering/templates/class-member.tsx [hunk 5]
```diff
@@ -75,7 +86,7 @@ export function ClassMember(props: {members: MemberEntryRenderable[]}) {
 
 function getMemberType(entry: MemberEntryRenderable): string | null {
   if (isClassMethodEntry(entry)) {
-    return entry.returnType;
+    return entry.implementation!.returnType;
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/rendering/templates/class-member.tsx:89]: Nit: Add comment on why `implementation` is never null here; or is this brittle?

File: adev/shared-docs/pipeline/api-gen/rendering/transforms/code-transforms.ts [hunk 6]
```diff
@@ -157,9 +163,10 @@ export function mapDocEntryToCode(entry: DocEntry): CodeTableOfContentsData {
         initialMetadata,
       );
     }
+
     return {
       // It is important to add the function keyword as shiki will only highlight valid ts
-      contents: `function ${getMethodCodeLine(entry, [], true)}`,
+      contents: `function ${getMethodCodeLine(entry.implementation!, [], true)}`,
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/rendering/transforms/code-transforms.ts:169]: These non-null casts seem brittle. Is there a way around this?

File: packages/compiler-cli/src/ngtsc/docs/src/function_extractor.ts [hunk 7]
```diff
@@ -106,3 +85,62 @@ export function extractAllParams(
...
+      ts.isCallSignatureDeclaration(decl) ||
+      ts.isMethodDeclaration(decl)
+    ) {
+      result.push({signature, decl});
+    }
+  }
+  return result;
+}
+
+export function extractOverloadSignatures(
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/docs/src/function_extractor.ts:108]:
```suggestion
export function extractCallSignatures(
```

File: packages/compiler-cli/src/ngtsc/docs/src/function_extractor.ts [hunk 8]
```diff
@@ -106,3 +85,62 @@ export function extractAllParams(
     isRestParam: !!param.dotDotDotToken,
   }));
 }
+
+/** Filters the list signatures to valid initializer API signatures. */
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/docs/src/function_extractor.ts:89]: Nit: copy leftover from the initializer API extraction

File: adev/shared-docs/pipeline/api-gen/rendering/templates/class-member.tsx (hunk 3 above)
Comment (JeanMeche) [reply adev/shared-docs/pipeline/api-gen/rendering/templates/class-member.tsx:46]: The condition is kind of included in `hasSignature` but I need to sort out the member/signature issue (your following comment), there are currently duplicate entries in the datastructure.

File: adev/shared-docs/pipeline/api-gen/rendering/shiki/shiki.ts (hunk 2 above)
Comment (JeanMeche) [reply adev/shared-docs/pipeline/api-gen/rendering/shiki/shiki.ts:52]: I extracted it into a function and made it an optional parameter.

File: adev/shared-docs/pipeline/api-gen/rendering/templates/class-member.tsx (hunk 4 above)
Comment (JeanMeche) [reply adev/shared-docs/pipeline/api-gen/rendering/templates/class-member.tsx]: I've entirely refactor this part.

File: adev/shared-docs/pipeline/api-gen/rendering/shiki/shiki.ts [hunk 9]
```diff
@@ -21,8 +21,12 @@ export async function initHighlighter() {
   });
 }
 
-export function codeToHtml(code: string, language: string | undefined): string {
-  return highlighter.codeToHtml(code, {
+export function codeToHtml(
+  code: string,
+  language: string | undefined,
+  removeFunctionKeyword = false,
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/rendering/shiki/shiki.ts:27]: Nit: An option object avoids the "boolean flag pattern".

File: adev/shared-docs/pipeline/api-gen/rendering/templates/highlight-ts.tsx [hunk 10]
```diff
@@ -6,15 +6,13 @@
...
-import { RawHtml } from './raw-html';
-import { codeToHtml } from '../shiki/shiki';
+import {h} from 'preact';
+import {RawHtml} from './raw-html';
+import {codeToHtml} from '../shiki/shiki';
 
 /** Component to render a header of the CLI page. */
 export function HighlightTypeScript(props: {code: string}) {
-  const result = codeToHtml(props.code, 'typescript');
+  const result = codeToHtml(props.code, 'typescript', /* removeFunctionKeyword */ true);
```
Comment (devversion) [inline adev/shared-docs/pipeline/api-gen/rendering/templates/highlight-ts.tsx:15]: Should we pass through the flag here as well? `highlight-ts` is pretty generic as well IMO

**Review by devversion — APPROVED** [review]
lgtm

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 04911d7599c3a410bdea56c72c4cc317a15ffc78.

The changes were merged into the following branches: main

**Pull request closed without merging**